### DISCUSSION TITLE
Bluetooth adapter not found in docker container

### DISCUSSION THREAD

**Discussion by Ripper346** [hbldh/bleak#1307] [Q&A] [answered]
* bleak version: 0.20.2
* Python version: 3.10
* Operating System: Ubuntu arm64, Linux kernel 5.16.17
* BlueZ version (`bluetoothctl -v`) in case of Linux: 5.55

### Description

I want to use bleak to scan Bluetooth beacons from a Docker container. I am coming from a bluepy installation that worked without configure anything, except for disabling bluetooth from the operating system, otherwise the resource would result occupied. When I start the container, bleak tells me that it can't find any adapter with this error:

```
Traceback (most recent call last):
  File "/executioner/raspberry-station/./main.py", line 51, in <module>
    asyncio.run(main())
  File "/usr/local/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/usr/local/lib/python3.10/asyncio/base_events.py", line 649, in run_until_complete
    return future.result()
  File "/executioner/raspberry-station/./main.py", line 47, in main
    await scanner_runner()
  File "/executioner/raspberry-station/scanner.py", line 117, in scanner_runner
    async with BleakScanner(lambda dev, adv: delegate.handleDiscovery(dev, adv)) as scanner:
  File "/root/.local/lib/python3.10/site-packages/bleak/__init__.py", line 152, in __aenter__
    await self._backend.start()
  File "/root/.local/lib/python3.10/site-packages/bleak/backends/bluezdbus/scanner.py", line 179, in start
    adapter_path = manager.get_default_adapter()
  File "/root/.local/lib/python3.10/site-packages/bleak/backends/bluezdbus/manager.py", line 299, in get_default_adapter
    raise BleakError("No Bluetooth adapters found.")
bleak.exc.BleakError: No Bluetooth adapters found.
```

### What I Did

So, at startup, as I said I close the OS bluetooth service with `killall -9 bluetoothd`.

Then I have a Docker compose for the services that I am using, in particular for the scanner I wrote
```
version: "3.9"

services:
  scanner_bl:
    image: beacon_station:4.1
    container_name: scanner_bl
    command: --job scanner
    environment:
      TZ: UTC
      DEFAULT_BRANCH: bleak
    volumes:
      - ./storage:/storage
    network_mode: "host"
    privileged: true
    restart: always
    depends_on:
      - db
```

The dockerfile I am using is this:
```
FROM python:3.10-slim-bullseye AS base
RUN apt-get update && \
    apt-get install -y git bluez bluetooth libpq5 iputils-ping

# compile
FROM base AS compile-image
RUN apt-get update && \
    apt-get install -y gcc make libffi-dev libgtk2.0-dev \
    postgresql-server-dev-all musl-dev
COPY requirements.txt /tmp/requirements.txt
RUN pip3 install --user -r /tmp/requirements.txt

# runtime
FROM base AS runtime-image
COPY --from=compile-image /root/.local /root/.local
ENV PATH=/root/.local/bin:$PATH
COPY src /src
WORKDIR /src
ENTRYPOINT ["./entrypoint.sh"]
```

With entrypoint.sh
```bash
#!/usr/bin/env bash

ARGS="$@"
service dbus start
bluetoothd &
python3 -u ./main.py $ARGS
```

My python code is at the moment taken from the documentation:
```python
import asyncio
from bleak import BleakScanner

async def scanner_runner():
    stop_event = asyncio.Event()
    scan_time = 15
    for i in range(0, 20 * 60, scan_time):
        async with BleakScanner(lambda dev, adv: print(dev, adv)) as scanner:
            await stop_event.wait(scan_time)

asyncio.run(scanner_runner())
```

At the very beginning of the container I am also calling from python the process `hciconfig` to see which is the mac address of my bluetooth card and it is correctly showing. Note as I said that I took straight the code and container from the code I had with bluepy (that it also uses bluez) and it was working.

Am I missing something or is it incompatible? Thank you.

**Comment by dlech** [comment]
What happens if you run `bluetoothctl` in the container?

**Reply by Ripper346** [reply]
So, first sorry for opening an issue and not a discussion topic, I am always missing them on github when they are enabled.

In the container that command opens a shell, it shows:
```
Agent registered
[bluetooth]#
```
It answers to the `version` with 5.55, if I go with `devices` it answers with `No default controller available`

**Reply by Ripper346** [reply]
I forgot to mention in my first post, in the entrypoint I have the enable of bluetooth:
```bash
#!/usr/bin/env bash

ARGS="$@"
service dbus start
bluetoothd &
python3 -u ./main.py $ARGS
```

**Reply by dlech** [reply]
If it can't find the adapter either, then you probably missed something in the Docker configuration.

A web search for "running bluez in docker container" shows lots of results so hopefully one of those can help.

**Reply by Ripper346** [reply]
I will take a look more deeply but the strange thing is that bluepy also uses bluez and it has no problems to use the bluetooth card from the container with the same configuration

**Reply by Ripper346** [reply]
Also, from the container if I run `hcitool dev` it answer with the card.
```
Devices:
        hci0   [mac]
```

**Comment by Ripper346** [comment, marked as answer]
So, apparently using `killall -9 bluetoothd` doesn't do much, it kills the bluetooth service but then for a Raspberry card the service is triggered again to open. This caused the bluetooth card to be busy and so being invisible for bluez inside the docker container. I was able to disable the bluetooth service in the OS, and bluez and consequently bleak, started working. Sorry for the trouble, in the end it wasn't even related to bleak
